contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,304 @@
+
+Using celery required us to build a Node app that:
+
+- Gets a stream of events from Python through a Redis queue (via [celery.node](https://celery-node.js.org/))
+- Runs user-defined JavaScript on that stream of events.
+- Sends them back to Python through the same queue for ingestion.
+
+The first and last steps were easy enough with our app structure and Celery, but we needed to make sure the arbitrary, user-defined JavaScript code was securely run. 
+
+### VMs to the rescue
+
+The solution to the second step was to use VMs (Virtual Machines). Turns out Node. v14 has a [built-in VM module](https://nodejs.org/docs/latest-v17.x/api/vm.html) that enables running custom JavaScript in a separate context.
+
+Using the VM module enabled us to run custom JavaScript code during event ingestion. The downside was that “the VM module is not a security mechanism. Do not use it to run untrusted code.” This wasn’t unexpected. [Privilege escalation](https://en.wikipedia.org/wiki/Privilege_escalation) and [resource exhaustion](https://en.wikipedia.org/wiki/Resource_exhaustion_attack) attacks are real. We couldn’t avoid them, but we could build strategies to mitigate them.
+
+Node's VM module puts your code in an isolated context, has ~~limited~~ no support for secure communications with the host, and has holes like this:
+
+```js
+const vm = require('vm');
+vm.runInNewContext('this.constructor.constructor("return process")().exit()');
+console.log('Never gets executed.');
+```
+
+Thus we needed an abstraction. The two most popular are `isolated-vm` and `vm2` :
+
+- [`isolated-vm`](https://github.com/laverdet/isolated-vm) claims to be secure and various big companies use it. Each "isolate" runs in a new thread, with controllable CPU and memory limits. There are methods to copy data between the main thread and an isolate, and we can share objects and functions between the host and the isolate. It's not a perfect sandbox, but it's as close as we can get.
+- [`vm2`](https://github.com/patriksimek/vm2) has a different isolation model. Each "VM" runs in an isolated NodeVM context, in the same thread as the rest of the app. There are no memory or CPU limits we can enforce. You run the code locally, but don’t share any variables with the host app.
+
+While `isolated-vm` felt like a great fit because of its emphasis on security, its implementation [wasn't a success](https://github.com/PostHog/posthog/issues/6855#issuecomment-853879421).
+
+> At the end of the day, I (Marius) would have had to implement some proxying code similar to vm2, just to get fetch working, and decided it's not worth the effort, considering the time budget and the extremely likely case that I'll leave in some security holes.
```

Comment:
Is this definitely needed? Kind of disrupts the flow. Could maybe do a footnote instead, or incorporate into the copy?